.github/workflows/testserver-helios.yml

```diff
@@ -0,0 +1,364 @@
+  validate-inputs:
+    runs-on: ubuntu-latest
+    steps:
+      - name: Validate Build Tag Input Exists
+        if: ${{ github.event.inputs.HELIOS_BUILD == 'true' && github.event.inputs.HELIOS_BUILD_TAG == '' }}
+        run: |
+          echo "HELIOS_BUILD is true but no HELIOS_BUILD_TAG was provided."
+          exit 1
+      - name: Validate PR Number Input Exists
+        if: ${{ github.event.inputs.HELIOS_BUILD == 'false' && github.event.inputs.HELIOS_PR_NUMBER == '' }}
+        run: |
+          echo "HELIOS_BUILD is false but no HELIOS_PR_NUMBER was provided."
+          exit 1
+
+      - name: Print Inputs
+        run: |
+          echo "RAW_URL: ${{ env.RAW_URL }}"
+          echo "Triggered by: ${{ github.event.inputs.HELIOS_TRIGGERED_BY }}"
+          echo "Branch: ${{ github.event.inputs.HELIOS_BRANCH_NAME }}"
+          echo "SHA: ${{ github.event.inputs.HELIOS_BRANCH_HEAD_SHA }}"
+          echo "PR Number: ${{ github.event.inputs.HELIOS_PR_NUMBER }}"
+          echo "Environment: ${{ github.event.inputs.HELIOS_ENVIRONMENT_NAME }}"
+          echo "Need Build? ${{ github.event.inputs.HELIOS_BUILD }}"
+          echo "Build Tag? ${{ github.event.inputs.HELIOS_BUILD_TAG }}"
+
+
+  # Build the Docker image (branch without PR) or check if the build has run successfully (PR)
+  build-or-check-existing-build:
+    needs: [ validate-inputs ]
+    runs-on: ubuntu-latest
+    steps:
+      # Build the Docker image (branch without PR)
+      - name: Git Checkout
+        if: ${{ github.event.inputs.HELIOS_BUILD == 'true' }}
+        uses: actions/checkout@v4
+        with:
+          ref: ${{ github.event.inputs.HELIOS_BRANCH_HEAD_SHA }}
+      - name: Set up QEMU
+        if: ${{ github.event.inputs.HELIOS_BUILD == 'true' }}
+        uses: docker/setup-qemu-action@v3
+      - name: Set up Docker Buildx
+        if: ${{ github.event.inputs.HELIOS_BUILD == 'true' }}
+        uses: docker/setup-buildx-action@v3
+      # Build and Push to GitHub Container Registry
+      - name: Login to GitHub Container Registry
+        if: ${{ github.event.inputs.HELIOS_BUILD == 'true' }}
+        uses: docker/login-action@v3
+        with:
+          registry: ghcr.io
+          username: ${{ github.repository_owner }}
+          password: ${{ secrets.GITHUB_TOKEN }}
+      - name: Build and Push to GitHub Container Registry
+        if: ${{ github.event.inputs.HELIOS_BUILD == 'true' }}
+        uses: docker/build-push-action@v5
+        with:
+          # beware that the linux/arm64 build from the registry is using an amd64 compiled .war file as
+          # the GitHub runners don't support arm64 and QEMU takes too long for emulating the build
+          platforms: linux/amd64,linux/arm64
+          file: ./docker/artemis/Dockerfile
+          context: .
+          tags: ghcr.io/ls1intum/artemis:${{ github.event.inputs.HELIOS_BUILD_TAG }}
+          push: true
+          cache-from: type=gha
+          cache-to: type=gha,mode=min
```

Comment:
This could be a separate job. That way you would only need one `if: ${{ github.event.inputs.HELIOS_BUILD == 'true' }}` and it would be better structured.

Reply:
Ideally you would also reuse the existing build workflow instead of copying it over. This could be done via reusable workflows.

Reply:
Also this will result in deployments that take an extra 10 minutes of runtime. During this time no other deployment can run. This is pretty bad and you should find a solution to this.